src/fs.ts

```diff
@@ -82,3 +74,95 @@ export class LocalFileSystem implements FileSystem {
 		return fs.readFile(this.resolveUriToPath(uri), 'utf8');
 	}
 }
+
+/**
+ * Memoization cache that saves URIs and searches parent directories too
+ */
+class ParentUriMemoizationCache extends Map<string | undefined, Promise<void>> {
+
+	/**
+	 * Returns the value if the given URI or a parent directory of the URI is in the cache
+	 */
+	get(uri: string | undefined): Promise<void> | undefined {
+		let hit = super.get(uri);
+		if (hit) {
+			return hit;
+		}
+		// Find out if parent folder is being fetched already
+		hit = super.get(undefined);
+		if (hit) {
+			return hit;
+		}
```

Comment:
`undefined` means no `base` was provided, in which case it defaults to the workspace root according to LSP